tensorflow/core/kernels/cwise_op_gpu_fma.cu.cc

```diff
@@ -0,0 +1,325 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+#include "tensorflow/core/kernels/cwise_op_fma.h"
+#include "tensorflow/core/kernels/cwise_ops_gpu_common.cu.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+template <FMAType Type, typename T> __device__ T fma_op(T m1, T m2)
+{
+    if (Type==FMAType_Add)
+      return m1 + m2;
+    else if (Type==FMAType_Sub)
+      return m1 - m2;
+    else
+      return m2 - m1;
+}
+
+//
+// Fast pathway: each tensor is either full-size or 1-element.
+//
+
+template <typename T, int N, FMAType Type>
```

Comment:
would it be cleaner to have the three bools as template params instead?
think that would make the code more readable and avoid the need to convert back and forth from `N`

something like the following here
```
template<typename T, bool broadcast_x1, bool broadcast_y1, bool broadcast_x2, FMAType Type>
...
```

and then call it like
```
   TF_CHECK_OK(GpuLaunchKernel(CwiseFusedMulAddKernel<T, broadcast_x1, broadcast_y1, broadcast_x2, Type>, 
...
```

Reply:
it'll still have to be a N at some point along the call stack, because of the indexing at lines 87 and 116. 
Without bools, I'll have to replace
`      &execute<0>,  &execute<1>,  &execute<2>,  &execute<3>,`

with
`      &execute<False,False,False,False>,  &execute<False,False,False,True>,  &execute<False,False,True,False>,  &execute<False,False,True,True>,
`
etc.

Reply:
can't you eliminate the  `execute` routine altogther?
i.e. wouldn't the following work
```
template<typename T, bool broadcast_x1, bool broadcast_y1, bool broadcast_x2, FMAType Type>
__global__ void CwiseFusedMulAddKernel(GpuLaunchConfig cfg, T* out, const T* x1,
                                       const T* y1, const T* x2) {
  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
    out[i]=fma_op<Type>(x1[broadcast_x1 ? 0 : i] * y1[broadcast_y1 ? 0 : i],
             x2[broadcast_x2 ? 0 : i]);
  }
}

template <typename T, FMAType Type>
void LaunchFusedMulAddOp<GPUDevice, T, Type>::operator()(
    const GPUDevice& device, T* out, const T* x1, const T* y1, const T* x2,
    uint64 elements, bool broadcast_x1, bool broadcast_y1, bool broadcast_x2) {
   auto config = GetGpuLaunchConfig(elements, device);
  TF_CHECK_OK(GpuLaunchKernel(CwiseFusedMulAddKernel<T, broadcast_x1, broadcast_y1, broadcast_x2, Type>,
                              config.block_count, config.thread_per_block, 0,
                              device.stream(), config, out, x1, y1, x2));
}
```

Reply:
No, it'll complain about broadcast_x1, etc. not known at compile time.

Reply:
that is strange...don't the template params need to be known at compile time by definition?

Reply:
Yes, which is why it will complain about passing a runtime argument of LaunchFusedMulAddOp as a template parameter.

Reply:
ah...got it